enacts/onset/agronomy.py

```diff
@@ -73,38 +76,95 @@ def soil_plant_water_balance(
+    sm, drainage, et_crop : Tuple of DataArray
+        daily soil moisture, drainage and crop evapotranspiration over the growing season.
         
     See Also
     --------
     soil_plant_water_step
     
+    Notes
+    -----
+    The daily evapotranspiration `et` can be scaled by a Crop Cultivar Kc
+    modelizing a crop needs in water according to the stage of its growth.
+    Kc is set to 1 outside of the growing period. i.e. before planting date
+    and after the last Kc curve inflexion point.
+    
+    Examples
+    --------
+    Example of kc_params:
+    
+    >>> kc_periods = pd.TimedeltaIndex([0, 45, 47, 45, 45], unit="D")
+    >>> kc_params = xr.DataArray(
+    >>>    data=[0.2, 0.4, 1.2, 1.2, 0.6], dims=["kc_periods"], coords=[kc_periods]
+    >>> )
+    <xarray.DataArray (kc_periods: 5)>
+    array([0.2, 0.4, 1.2, 1.2, 0.6])
+    Coordinates:
+        * kc_periods  (kc_periods) timedelta64[ns] 0 days 45 days ... 45 days 45 days
+    
+    Example of planting_date:
+    
+    >>> p_d = xr.DataArray(
+    >>>    pd.DatetimeIndex(data=["2000-05-02", "2000-05-13"]),
+    >>>    dims=["station"],
+    >>>    coords={"station": [0, 1]},
+    >>> )
+    <xarray.DataArray (station: 2)>
+    array(['2000-05-02T00:00:00.000000000', '2000-05-13T00:00:00.000000000'],
+        dtype='datetime64[ns]')
+    Coordinates:
+        * station        (station) int64 0 1
     """
     
     # First Step
     if np.size(et) == 1:
         et = xr.DataArray(et)
+    # Setting Kc
+    if kc_params is None:
+        kc = 1
+    else:
+        kc_inflex = kc_params.assign_coords(
+            kc_periods=kc_params["kc_periods"].cumsum(dim="kc_periods")
+        )
+        if planting_date is not None:
+            planted_since = peffective[time_dim][0].drop(time_dim) - planting_date
+            kc = kc_inflex.interp(
+                kc_periods=planted_since, kwargs={"fill_value": 1}
+            ).drop("kc_periods")
+    # Initializations of sm, drainage and et_crop
+    et_crop0 = kc * et.isel({time_dim: 0}, missing_dims='ignore', drop=True)
     sm0, drainage0 = soil_plant_water_step(
         sminit,
         peffective.isel({time_dim: 0}, drop=True),
-        et.isel({time_dim: 0}, missing_dims='ignore', drop=True),
+        et_crop0,
         taw,
     )
-    # Give time dimension to sm and drainage    
+    # Give time dimension to sm, drainage and et_crop
     sm = sm0.expand_dims({time_dim: peffective[time_dim]}).copy()
     drainage = drainage0.expand_dims({time_dim: peffective[time_dim]}).copy()
+    et_crop = et_crop0.expand_dims({time_dim: peffective[time_dim]}).copy()
     # Filling/emptying bucket day after day
     for doy in range(1, peffective[time_dim].size):
+        if kc_params is not None and planting_date is not None:
+            planted_since = planted_since + pd.Timedelta(days=1)
+            kc = kc_inflex.interp(kc_periods=planted_since, kwargs={"fill_value": 1})
```

Comment:
If I understand correctly what's happening here, it would be more efficient to calculate the entire kc vector outside the loop, rather than calculating one element of it on each iteration of the loop.

Reply:
Oh, you can't because planting_date is determined on the fly. Maybe once you find the planting date, then you calculate the entire kc vector?

Reply:
planting_date is not determine on the fly yet, but it will. Then the problem is that planting_date may vary with other dimensions, e.g. space, so it will still need be in the loop until all planting dates are found or something like that. Unless you see it otherwise.